lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -855,6 +855,101 @@ static void optimizeBatchNorm(Function *F) {
+  //  1). The inputs dims vectors are all the same. e.g. <2,2,3> ... <2,2,3>.
+  //      For this case, the returned dim should be the chosen index that can
+  //      maintain the data order as by the original ConcatNode (origConcatN).
+  //      A example is like, the origConcatN concatenates two <4,3> on the 2nd
+  //      dim, then the new Concat can only concatenate on the 2nd dim, as we
+  //      want the concat "granularity" to be exactly 3 (which is of the
+  //      origConcatN).
+  //  2). Only one dim is different for all the dims vectors. e.g. <2,1,3>,
+  //      <2,2,3>, <2,2,3>, <2,3,3>. For this case, the returned dim is this dim
+  //      i where dims1[i] != dims2[i]. While at the same time, we also need to
+  //      make sure the concat "granularity" to be exactly same as of the
+  //      origConcatN.
+  auto firstDims = inputs.front().dims();
+  size_t dimSize = firstDims.size();
+  auto origCNInputShape = origConcatN->getInputs().front().dims();
+  int numOfUnitToLegallyConcat = 1;
+  // Find the minimal number of data units so that we can legally Concat them.
+  for (size_t i = origConcatN->getDim() + 1; i < origCNInputShape.size(); ++i) {
+    numOfUnitToLegallyConcat *= origCNInputShape[i];
+  }
+
+  ssize_t index = -1;
+  ssize_t preIndex = -1;
+  // Go through all the input dims vectors, find out if either all vectors
+  // are all the same, or they are equal in all but one dim.
+  for (auto &in : inputs) {
+    auto curDims = in.dims();
+    // Find the previous non-negative index value.
+    if (index != -1) {
+      if (preIndex != -1 && preIndex != index) {
+        // We cannot have two different non-negative index to be the
+        // dim for Concat.
+        return -1;
+      }
+      preIndex = index;
+    }
+    index = -1;
+    for (size_t i = 0; i < dimSize; ++i) {
+      // Find the index i where firstDims[i] != curDims[i]
+      if (firstDims[i] != curDims[i]) {
+        if (index != -1) {
+          // If curDims and firstDims have more than one different element ,
+          // then we can not perform Concat on top of these two nodes.
+          return -1;
+        }
+        index = i;
+      }
+    }
+  }
+
+  // Find the non-negative index value.
+  if (index == -1) {
+    index = preIndex;
+  }
+  // It's still possible that index can be -1. it means all inputs have
+  // same shape.
+  if (index == -1) {
+    // In this case, we need to determine the index that can meet the
+    // requirement of the number of data units in the small tensors that would
+    // be concatenated.
+    int cumulativeNumUnits = 1;
+    for (ssize_t i = dimSize - 1; i >= 0; i--) {
+      // Go over the dimensions from right to left, find the first index that
+      // can make cumulativeNumUnits divided by numOfUnitToLegallyConcat.
+      if (cumulativeNumUnits == numOfUnitToLegallyConcat) {
+        return i;
+      }
+      cumulativeNumUnits *= firstDims[i];
+    }
+    return -1;
+  } else {
```

Comment:
No need for else here as the if-branch ends with a return.

Reply:
right. removed the else branch